### ISSUE TITLE
difficulties with overriding

### ISSUE THREAD

**Issue by TrejGun** [hoardexchange/hoard-exchange-erc998#14]
Hi guys, thanks for a good job!

We implement the game with erc998 tokens, and I just tried to extend my in-game item from your ComposableTopDown contract and here is my feedback.

We also use @OpenZepplin as our core, and our tradable abstract class looks like so

```
abstract contract ERC721Tradable is ERC721Capped, ERC721URIStorage, ERC721Pausable, ERC721Burnable, ComposableTopDown {
```

As you can see, everything extends ERC721, but ComposableTopDown extends IERC721 and implements its methods by itself. The problem is: I have to override each and every method from ERC721 to point to ComposableTopDown.

How do you think is there an easy way to work around this? Can I use another OpenZepplin presets together with Composable without overriding 25 methods? What if you extend ERC721 instead of IERC721? Will this reduce the number to 18?

**Comment by TrejGun** [comment]
I also can't override methods like isApprovedForAll and getApproved because it is not virtual. This makes it impossible to use your code with other presets

**Comment by TrejGun** [comment]
so far the biggest problem is _checkOnERC721Received which is private and not overridable. it looks like I will have to rewrite safeMint using methods from openzepplin as well as integrate enumerable preset

**Comment by TrejGun** [comment]
Hi guys! I have OpenZepplin compatible extension for 998, please PM

**Comment by lukasz-glen** [comment]
The contract cannot inherit `ERC721` instead of `IERC721` - implementations differ.
I changed visibility.
Since that time the contract has grown large, it is close to the limits.

**Comment by TrejGun** [comment]
btw I already have a working version of this extension with tests

**Comment by TrejGun** [comment]
Hey @lukasz-glen 

I see you did a great job on improving this contract. And I want to share my feedback with you one more time. A long long time ago I was in a team that developed this standard. Since that time I use this standard from time to time in my games. 

Consider the real-world applications, that most likely will not use contracts, not from its ecosystem. 

![valkadil1a](https://user-images.githubusercontent.com/185285/172875770-70a32783-6608-4f62-a8ca-5536b4b51b9c.jpg

Say I have a Diablo-like game where the character can wear some gears, this character is 998, the gears are 998 and/or 721 and gems/runes are 721. I most likely will not allow users to equip CryptoKitty as a helmet (there is no such texture in the game). This means I have to add a whitelist for contracts that could be used as children. You most likely will not allow users to wear 10 helmets at a time too, so you have to add a limitation here too. Adding this code will definitely reach mainnet limit which makes this contract not usable.

Another example. I have the same game and 1155 tokens which represent resources for crafting. I would like to implement my backpack to handle only 100kg of resources. More code...

Ok, let's try another game I have an army with squads and soldiers. 998+721 -> 998+1155 -> 1155 . I just don't need all the other extensions

All these examples a pretty flat. I mean I can't move gem/rune directly to the character's hands. so BU is kinda redundant. 

Then I definitely need access control or an ownable extension to sell my items from my marketplace. Speaking of marketplaces: OpenSea uses a trading protocol which executes meta-transactions so msg.sender would be invalid. Why don't you use context?

In conclusion, I want to say that this contract is bloated and should be split into core and several small extensions same as OZ does. I know you did a great job and most likely have a bigger picture of your ecosystem but in the end, you are doing this for your customers with their use cases and we will appreciate it if you can make it flexible

**Comment by lukasz-glen** [comment]
@TrejGun 
Thanks for your comments. Sorry for the late reply, I've got a lot of work recently. 
Thanks for pointing out the lack of Context. Natural.
I like the idea to decompose the contract, OZ style. For sure I could make erc1155 support and old erc721 (CK) support optional. Not sure if erc721 can be easily made optional because of rootOwnerOf() and other functions that operates on a tree.
Our implementation have also state-hash which is a kind of trade protection. I will have to think a little how to move it to an extension. 
And my other thought was that Enumerables generate a lot of code and cost gas. So maybe those should be extensions also. I mean, dApps can use indexers or read events themselves to track contents of composables. What is your opinion?
I will put my hands on it after the weekend.

**Comment by lukasz-glen** [comment]
A little step forward ^^. You can take a look if you have time.

**Comment by TrejGun** [comment]
oh wow) that was fast
from the first look direct replacement of msg.sender for _msgSender() will work but you can save some gas by assigning it to variable

**Comment by lukasz-glen** [comment]
You may want to look at this ^^, in particular how erc20 support was decomposed. There are `_beforeSmth()` and `_afterSmth()` functions like OZ has. Maybe these will be helpful for you.

**Comment by TrejGun** [comment]
Oh I really like those functions)

I did not try to use the new code yet but I tried the old one and want to report one more issue
there are several checks like `tokenId != 0` like this one

https://github.com/hoardexchange/hoard-exchange-erc998/blob/main/contracts/ComposableTopDown.sol#L530

I understand how you come to this condition - the initial value of uint256 is 0, so you guess it is initial, but in fact, 0 is a valid token id. So there are two ways to fix this, to call `increment` immediately 
```sol
  constructor() {
      _tokenIdTracker.increment();
  }
```

or probably a less secure way to switch lines (OZ does increment first)
```
  function mint(address to) public virtual override onlyRole(MINTER_ROLE) {
    _mint(to, _tokenIdTracker.current());
    _tokenIdTracker.increment();
  }
```

not having 0 tokens is totally fine with me, because this way it is aligned with the sequence in the database (Postgres) but may cause problems to others

**Comment by TrejGun** [comment]
Completely forgotten by the time of the last message but definitely worth mentioning.
It would be good to have a function that strips tokens, a cost-efficient way to remove all children in one transaction. 
Should not be a big security concern as all tokens are transferred to the same wallet
Here is an example from OpenSea
https://github.com/ProjectOpenSea/seaport/blob/main/contracts/lib/TokenTransferrer.sol
but for me, this is pure magic

**Comment by lukasz-glen** [comment]
I briefly checked 0 token id issue. It is only about composables. Zero id nfts are supported as children. The problem is this and similar statements: `uint256 tokenId = childTokenOwner[_childContract][_childTokenId];`. And you cannot say whether `childContract.childTokenId` belongs to zero `tokenId` or does not belong to any composable. For now I just added a simple check in `_mint`. I will have to spend more time to support zero id composables.

There are two concerns about stripping functions. Unpredictable gas consumption - a call can run out of gas. And enumerations are required because a function has to know the content of composables. Maybe batch operations would be enough?
The link you sent is how to call transfers (transferFrom) in gas efficient manner. I guess there are some assumptions though. Anyway, two concerns above remains.

**Comment by TrejGun** [comment]
As I said we do not use tokenId 0 ourselves but this is a potential pitfall so I told you about this

Can you please clarify the difference between the striping function and batch operations?

**Comment by lukasz-glen** [comment]
I separated `ComposableTopDownERC1155Enumerable` and `ComposableTopDownERC1155` has OZ-like functions.

I imagine that a stripping function is `bundleContract.empty(tokenId, receiver)`. With a batch function you have to pass content as arguments, for instance `bundleContract.transferBatchERC20(tokenId, receiver, erc20ContractsArray, amountsArray)`. In the the first case `bundleContract` has to know its content, in the second case it is just a batch operation.

**Comment by TrejGun** [comment]
you probably can do it the same way as you did with _beforeRemove. So new _beforeEmpty will just emit an event and then it would be overridden in each extension with another function that contains the logic

**Comment by TrejGun** [comment]
Hi @lukasz-glen 
I was working around your contracts and want to share a small fix https://github.com/hoardexchange/hoard-exchange-erc998/pull/23
